### PR TITLE
NIFI-4914

### PR DESCRIPTION
**Pull request by david-streamlio**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by lfrancke** [comment]
The pull request title and the commit is missing the JIRA number, could you add that?

>  Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

**Comment by david-streamlio** [comment]
Any update on this? I don't want to miss another release window if possible... =)

**Comment by joewitt** [comment]
sorry @david-streamlio  i def would like to get it in as well.  i will try to help soon

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    public static final PropertyDescriptor PULSAR_SERVICE_URL = new PropertyDescriptor
+            .Builder().name("PULSAR_SERVICE_URL")
+            .displayName("Pulsar Service URL")
+            .description("URL for the Pulsar cluster, e.g localhost:6650")
+            .required(true)
+            .addValidator(StandardValidators.HOSTNAME_PORT_LIST_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
+
+    public static final PropertyDescriptor ALLOW_TLS_INSECURE_CONNECTION = new PropertyDescriptor.Builder()
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java]: This probably should not be an option and the security context service should encapsulate such things.

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    public static final PropertyDescriptor STATS_INTERVAL = new PropertyDescriptor.Builder()
+            .name("Stats interval")
+            .description("The interval between each stat infomation update. It should be set to at least 1 second")
+            .required(false)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .defaultValue("60 sec")
+            .build();
+
+    public static final PropertyDescriptor TLS_TRUST_CERTS_FILE_PATH = new PropertyDescriptor.Builder()
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java]: if the SSL Context Service is being passed in this should not be necessary.

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java (hunk 1 above)
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java]: This property allows the Pulsar client to accept an untrusted TLS certificate from broker, which may be an edge use case, but is still a situation that the Pulsar client API exposes, and I want to enable users to configure the client as they would in any other situation.

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java (hunk 2 above)
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java]: The SSL Context service only encapsulates two value, a KeyStore file and a TrustStore file, which I use to configure the broker endpoint. But there is also another property used to set the path to the trusted TLS certificate file (for the X509 certificate, etc) the broker uses, so I had to expose that setting in a separate property.

https://pulsar.incubator.apache.org/api/client/ has the detailed explanations for these properties.

**Comment by david-streamlio** [comment]
Any update on this? Am I safe to assume we are going to make the 1.8 release?

**Comment by joewitt** [comment]
not clear when we'll be cutting a 1.8 release so i'm not sure about safety

but this is clearly a cool capability and it is just a matter of finding a committer to review it with sufficient bandwidth and expertise. things impacting security are super important and not being record oriented makes it less useful for sure.  I haven't looked at the details in a while to see if you added that.  I'd go so far as to recommend not offering a non record approach but i wouldnt say that is a rule - just a recommendation

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            "Discard the message and don't perform any addtional processing on the message");
+    static final AllowableValue FAIL = new AllowableValue(ConsumerCryptoFailureAction.FAIL.name(), "Fail",
+            "Report a failure condition, and the route the message contents to the FAILED relationship.");
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("FlowFiles for which all content was consumed from Pulsar.")
+            .build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
```
[the remark this answers is not in the document]
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java]: Good point. I have moved this relationship to the ConsumePulsarRecord class, which will route data that it receives, but cannot read with the configured reader to this relationship.

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    static final AllowableValue SHARED = new AllowableValue("Shared", "Shared", "Multiple consumer will be able to use the same subscription name and the messages");
+    static final AllowableValue FAILOVER = new AllowableValue("Failover", "Failover", "Multiple consumer will be able to use the same subscription name but only 1 consumer "
+                    + "will receive the messages. If that consumer disconnects, one of the other connected consumers will start receiving messages");
+
+    static final AllowableValue CONSUME = new AllowableValue(ConsumerCryptoFailureAction.CONSUME.name(), "Consume",
+            "Mark the message as consumed despite being unable to decrypt the contents");
+    static final AllowableValue DISCARD = new AllowableValue(ConsumerCryptoFailureAction.DISCARD.name(), "Discard",
+            "Discard the message and don't perform any addtional processing on the message");
+    static final AllowableValue FAIL = new AllowableValue(ConsumerCryptoFailureAction.FAIL.name(), "Fail",
+            "Report a failure condition, and the route the message contents to the FAILED relationship.");
```
[the remark this answers is not in the document]
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java]: Fixed

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/pom.xml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-client-service-api</artifactId>
+            <version>1.8.0-SNAPSHOT</version>
+            <scope>provided</scope>
+	    </dependency>
+        <dependency>
+            <groupId>org.apache.pulsar</groupId>
+            <artifactId>pulsar-client</artifactId>
+            <version>2.0.0-rc1-incubating</version>
```
[the remark this answers is not in the document]
Comment (david-streamlio) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/pom.xml]: Fixed

**Comment by pvillard31** [comment]
Hi @david-streamlio - sorry it took so long, I've been busy with a lot of things...

I've not been able to complete all my tests successfully but I wanted to give a status with my tests so far. All the code changes I did are available here: 8ca2db6712b9fc843a3d2ee7917ecc6db2dd54f2 Most of the changes are just to be consistent with the code base but some have been required to make things work.

In addition to that, I have the following comments: if we leave the ack timeout to 0sec, we get the following error

````
2018-10-22 08:23:18,998 WARN [Timer-Driven Process Thread-8] o.a.n.controller.tasks.ConnectableTask Administratively Yielding ConsumePulsar[id=1a70a923-0166-1000-e698-63a0160cc922] due to uncaught Exception: java.lang.IllegalArgumentException: Ack timeout should be should be greater than 1000 ms
java.lang.IllegalArgumentException: Ack timeout should be should be greater than 1000 ms
at org.apache.pulsar.shade.com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
at org.apache.pulsar.client.impl.ConsumerBuilderImpl.ackTimeout(ConsumerBuilderImpl.java:147)
at org.apache.nifi.processors.pulsar.AbstractPulsarConsumerProcessor.getConsumerBulder(AbstractPulsarConsumerProcessor.java:361)
at org.apache.nifi.processors.pulsar.AbstractPulsarConsumerProcessor.getConsumer(AbstractPulsarConsumerProcessor.java:338)
at org.apache.nifi.processors.pulsar.pubsub.ConsumePulsar.onTrigger(ConsumePulsar.java:46)
at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1165)
at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:203)
at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:117)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
````

Is it something that changed with the latest version of Pulsar?

In the consume processors, I'd add the topic name as attribute of flowfile and possibly other attributes that could be useful for later routing/filtering.

When stopping the processors or stopping NiFi, some processors does not stop cleanly and it requires a process kill for NiFi to shutdown. During my last test, it was on ConsumePulsar but it could be on other processors as well.

**Comment by david-streamlio** [comment]
@pvillard31 Thanks for taking the time to review my PR.  

1. How should I go about applying your changes to my branch?  Would "git cherry-pick 8ca2db6" work?

2. Regarding the " the ack timeout to 0sec", yes that is a new change in Pulsar. So we will need to make that property required.

3. Great suggestion on the consume processors attributes. I will review these and add some more of these for the downstream processors to use.

4. Re: "stopping the processors or stopping NiFi".  I have noticed that behavior as well and am not 100% sure what is causing it. AFAIK, on shutdown I am explicitly stopping all of the running threads in the executor service, waiting a few seconds for them to complete and then doing a "shutdown now".  Any ideas on what might be causing the issue?  Have you seen this behavior before in other processors?

**Comment by rumbin** [comment]
Is there an estimation when this may be merged? I'm also very much interested in using Pulsar within NiFi, so I can provide some application testing, if needed...

**Comment by david-streamlio** [comment]
@rumbin I am merging in changes from @pvillard31, and correcting some minor issues that cause the processors to hang in certain situations. ETA is by Monday 11/12 for my next commit with these changes

**Comment by joewitt** [comment]
travis and local builds dont work due to some parent relative path issues.  can you please confirm you're following the pattern that other components follow with regards to pom contents

**Comment by david-streamlio** [comment]
Thanks @joewitt I was double comparing the nifi-pulsar-bundle/pom.xml to other bundles in the nifi-nar-bundles, e.g. nifi-redis-bundle/pom.xml, nifi-rethinkdb-bundle/pom.xml, and nidi-parquet-bundle/pom.xml and they all define their parent module as follows:

    <parent>
        <groupId>org.apache.nifi</groupId>
        <artifactId>nifi-parquet-bundle</artifactId>
        <version>1.8.0-SNAPSHOT</version>
    </parent>

Which is what I have as the definition in the nifi-pulsar-bundle/pom.xml file

**Comment by joewitt** [comment]
the build doesn't work...at least for me or travis.  also not with the new patch.  i've not looked into why yet but we dont use relative paths anywhere that i know of so not sure what the latest change was for/doing

**Comment by david-streamlio** [comment]
@joewitt So the build doesn't work for you locally?  I am able to build it locally. 

 The latest commit was  trying suggestion number 2 from the following wiki page that is given in the build output: 
https://cwiki.apache.org//confluence/display/MAVEN/ProjectBuildingException
 https://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException

"You intent to use a parent POM from the local filesystem but Maven didn't use that. Please verify that the <relativePath> element in the child is properly set and that the POM at that location has actually the version you want to use."

**Comment by david-streamlio** [comment]
As a sanity check I did the following on my local MacBook, and was able to build NiFi successfully from my branch:

cd /tmp
git clone git@github.com:david-streamlio/nifi.git
cd nifi/
git checkout NIFI-4914
mvn clean install -DskipTests

...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 09:48 min
[INFO] Finished at: 2018-11-16T12:45:01-08:00
[INFO] Final Memory: 318M/1815M
[INFO] ------------------------------------------------------------------------

**Comment by joewitt** [comment]
am building against master - is 1.9.0.  changing it to 1.9.0-SNAPSHOT and ignoring last commit addresses it.  travis fails for a similar reason i'd guess.  Probably makes sense to ditch the merge commits, squash, rebase to latest.  But in any case is building now i think

**Comment by david-streamlio** [comment]
Is there a way to get this into a 1.8.x release rather than the 1.9.x?

**Comment by rumbin** [comment]
@david-streamlio why did you close it? Will you submit a new PR?

**Comment by david-streamlio** [comment]
@rumbin I am going to submit a new PR, it is just easier to create a fresh branch than rebasing to a new version of NiFi.

**Comment by joewitt** [comment]
@david-streamlio its not likely we'll put a feature like this into an already released line.  It doesn't fit our semver scheme.

Also, on my local branch i have the changes i needed from your last commit so this would actually build against master.  However, if I do parallel builds the build gets hung up.  Single threaded builds are fine.  I cannot tell why yet.  If I do a full clean build parallel just on master things are fine.

So, i believe there is something in the way these tests run that break parallel builds

File: nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java (hunk 1 above)
Comment (joewitt) [reply nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java]: This property should be removed as its meaning is too vague.  There are a lot of ways a TLS connection could be considered or made insecure.  If you mean to allow for untrusted certs, and I really dont understand why we would, then it should be clear to what it is doing.

**Comment by david-streamlio** [comment]
Created new PR after rebasing to NiFi 1.9.0-Snapshot branch. 

https://github.com/apache/nifi/pull/3178

**Pull request closed without merging**